<a href="https://colab.research.google.com/github/sohanchatterjee/ISE-490/blob/main/Copy_of_EM490_HW4_Pandas_Starter.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Please enter your name and student ID here.**

Sohan Chatterjee

20006064

**Homework - Pandas**

**Use Pandas to solve questions 1-5.**

In [25]:
# Mount your drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [12]:
# Change current directory
%cd /content/drive/My\ Drive/Colab\ Notebooks

/content/drive/My Drive/Colab Notebooks


In [13]:
 # Verify the current directory
%pwd

'/content/drive/My Drive/Colab Notebooks'

In [26]:
 # Verify the current directory
%ls

'Copy of Convert ipynb to HTML in Colab.ipynb'        'Copy of PythonBasics_Starter.ipynb'
'Copy of EM490- HW2_RegularExpression_Starter.ipynb'   EmissionsData.csv
'Copy of EM490- HW3_NumPy_Starter.ipynb'               ISE490-HW1_PythonBasics_Starter.ipynb
'Copy of EM490-HW4_Pandas_Starter.ipynb'


## Q1. (0.5 points)
Import the Pandas package and load the data from **EmissionsData.csv** into a dataframe called **em_df**.

In [15]:
# Provide your answer here.
import pandas as pd

em_df = pd.read_csv("EmissionsData.csv")

## Q2. (1 point)

Sort the dataframe using the `sort_values()` function by the column '**Emission**' in descending order. Display the top 5 records.

<div style="overflow-x:auto;">
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Year</th>
      <th>Country</th>
      <th>Continent</th>
      <th>Emission</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>592</th>
      <td>2008</td>
      <td>Qatar</td>
      <td>Asia</td>
      <td>48.601620</td>
    </tr>
    <tr>
      <th>593</th>
      <td>2009</td>
      <td>Qatar</td>
      <td>Asia</td>
      <td>44.836401</td>
    </tr>
    <tr>
      <th>595</th>
      <td>2011</td>
      <td>Qatar</td>
      <td>Asia</td>
      <td>44.018926</td>
    </tr>
    <tr>
      <th>594</th>
      <td>2010</td>
      <td>Qatar</td>
      <td>Asia</td>
      <td>42.639076</td>
    </tr>
    <tr>
      <th>710</th>
      <td>2010</td>
      <td>Trinidad And Tobago</td>
      <td>North America</td>
      <td>38.337841</td>
    </tr>
  </tbody>
</table>
</div>

In [17]:
# Provide your answer here.
em_df_sorted = em_df.sort_values(by='Emission', ascending=False)
em_df_sorted.head(5)

Year              Country      Continent   Emission
592  2008                Qatar           Asia  48.601620
593  2009                Qatar           Asia  44.836401
595  2011                Qatar           Asia  44.018926
594  2010                Qatar           Asia  42.639076
710  2010  Trinidad And Tobago  North America  38.337841

## Q3.

### Q3.1 (1 point)

Use the `groupby()` function to group **em_df** by the '**Year**' column..

In [18]:
# Provide your answer here.
em_df_grouped = em_df.groupby('Year')

### Q3.2 (1.5 points)

Retrieve the group '**2011**' using the `get_group()` function, and sort the 2011 group dataframe by '**Emission**' in ascending order. Display the first 5 rows.


In [19]:
# Provide your answer here.
em_2011 = em_df_grouped.get_group(2011)
em_2011_sorted = em_2011.sort_values(by='Emission', ascending=True)
em_2011_sorted.head(5)

Year           Country Continent  Emission
51   2011           Burundi    Africa  0.021350
683  2011              Chad    Africa  0.043830
779  2011  Dem Rep Of Congo    Africa  0.050303
639  2011           Somalia    Africa  0.058707
607  2011            Rwanda    Africa  0.062874

### Q3.3 (1 point)

Use the `mean()` function to calculate the average emission of all countries in the year **2011**.

<div style="overflow-x:auto;">
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Year</th>
      <th>Country</th>
      <th>Continent</th>
      <th>Emission</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>51</th>
      <td>2011</td>
      <td>Burundi</td>
      <td>Africa</td>
      <td>0.021350</td>
    </tr>
    <tr>
      <th>683</th>
      <td>2011</td>
      <td>Chad</td>
      <td>Africa</td>
      <td>0.043830</td>
    </tr>
    <tr>
      <th>779</th>
      <td>2011</td>
      <td>Dem Rep Of Congo</td>
      <td>Africa</td>
      <td>0.050303</td>
    </tr>
    <tr>
      <th>639</th>
      <td>2011</td>
      <td>Somalia</td>
      <td>Africa</td>
      <td>0.058707</td>
    </tr>
    <tr>
      <th>607</th>
      <td>2011</td>
      <td>Rwanda</td>
      <td>Africa</td>
      <td>0.062874</td>
    </tr>
  </tbody>
</table>
</div>

```text
Average: 4.904900274527919
```

In [20]:
# Provide your answer here.
print("Average: ", em_2011['Emission'].mean())

Average:  4.904900274527918


## Q4.

### Q4.1 (2.0 points)

Find the rows in `em_df` that meet the following conditions: **Year is 2010, Continent is Europe, and Emission is greater than 2**. Sort the resulting dataframe by '**Emission**' in descending order, and display the first 5 rows.


In [21]:
# Provide your answer here.
filtered_em_df = em_df[(em_df['Year']==2010)&(em_df['Continent']=='Europe')&(em_df['Emission']>2)]
filtered_em_df_sorted = filtered_em_df.sort_values(by='Emission', ascending=False)
filtered_em_df_sorted.head(5)

Year        Country Continent   Emission
434  2010     Luxembourg    Europe  21.635136
226  2010        Estonia    Europe  13.773197
246  2010  Faroe Islands    Europe  12.986678
534  2010         Norway    Europe  11.616193
234  2010        Finland    Europe  11.530840

### Q4.2 (1 point)

Calculate the average emission of **the resulting dataframe from the previous conditions**.

<div style="overflow-x:auto;">
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Year</th>
      <th>Country</th>
      <th>Continent</th>
      <th>Emission</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>434</th>
      <td>2010</td>
      <td>Luxembourg</td>
      <td>Europe</td>
      <td>21.635136</td>
    </tr>
    <tr>
      <th>226</th>
      <td>2010</td>
      <td>Estonia</td>
      <td>Europe</td>
      <td>13.773197</td>
    </tr>
    <tr>
      <th>246</th>
      <td>2010</td>
      <td>Faroe Islands</td>
      <td>Europe</td>
      <td>12.986678</td>
    </tr>
    <tr>
      <th>534</th>
      <td>2010</td>
      <td>Norway</td>
      <td>Europe</td>
      <td>11.616193</td>
    </tr>
    <tr>
      <th>234</th>
      <td>2010</td>
      <td>Finland</td>
      <td>Europe</td>
      <td>11.530840</td>
    </tr>
  </tbody>
</table>
</div>

```text
Average: 7.4887038956
```

In [22]:
# Provide your answer here.
print("Average: ", filtered_em_df_sorted['Emission'].mean())

Average:  7.4887038956


## Q5.

### Q5.1 (1 point)

In **em_df**, create a new column called '**Emission10PMore**'. The value in this column should be 10% more than the corresponding value in the original '**Emission**' column.


In [23]:
# Provide your answer here.
em_df['Emission10PMore'] = em_df['Emission']*1.10

### Q5.2 (1 point)

Sort the resulting dataframe by '**Emission10PMore**' in ascending order and display the first 5 rows.

<div style="overflow-x:auto;">
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Year</th>
      <th>Country</th>
      <th>Continent</th>
      <th>Emission</th>
      <th>Emission10PMore</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>50</th>
      <td>2010</td>
      <td>Burundi</td>
      <td>Africa</td>
      <td>0.020542</td>
      <td>0.022596</td>
    </tr>
    <tr>
      <th>49</th>
      <td>2009</td>
      <td>Burundi</td>
      <td>Africa</td>
      <td>0.020868</td>
      <td>0.022954</td>
    </tr>
    <tr>
      <th>51</th>
      <td>2011</td>
      <td>Burundi</td>
      <td>Africa</td>
      <td>0.021350</td>
      <td>0.023485</td>
    </tr>
    <tr>
      <th>48</th>
      <td>2008</td>
      <td>Burundi</td>
      <td>Africa</td>
      <td>0.021615</td>
      <td>0.023777</td>
    </tr>
    <tr>
      <th>681</th>
      <td>2009</td>
      <td>Chad</td>
      <td>Africa</td>
      <td>0.042689</td>
      <td>0.046958</td>
    </tr>
  </tbody>
</table>
</div>


In [24]:
# Provide your answer here.
em_df_10p_sorted = em_df.sort_values(by='Emission', ascending=True)
em_df_10p_sorted.head(5)

Year  Country Continent  Emission  Emission10PMore
50   2010  Burundi    Africa  0.020542         0.022596
49   2009  Burundi    Africa  0.020868         0.022954
51   2011  Burundi    Africa  0.021350         0.023485
48   2008  Burundi    Africa  0.021615         0.023777
681  2009     Chad    Africa  0.042689         0.046958